### ISSUE TITLE
"Waiting for decode worker"

### ISSUE THREAD

**Issue by sbradnick** [Xpra-org/xpra-html5#183]
This could be an issue w/ my expectations or plain old user error, but it seems like xpra-html5 isn't functional for me after 4.5.2+git20220421.ecc3a08 (basically git commit ecc3a08 on April 21st, 2022).

I'm running: `xpra start --start=xterm -d auth --bind-tcp=0.0.0.0:14500,auth=sys` and then firing up either firefox or qutebrowser to connect to the host.  All works fine with ecc3a08, but using either of these, it never seems to quite work:

5.0+git20220519.25a2b99-1.1 [25a2b99]
5.0+git20220523.59a8f90-1.1 [59a8f90]

I either get "Waiting for decoder worker" on a regular interval (maybe every 21s with 5.0+git20220523.59a8f90-1.1) or dropped back to the login page (with 5.0+git20220519.25a2b99-1.1).

```
2022-05-23 14:19:18,528 Error: connection timed out: ws socket: 192.168.0.231:14500 <- 192.168.0.106:51328
2022-05-23 14:19:18,564  after 21 seconds
2022-05-23 14:19:18,574  received 0 bytes
```

I'm attaching a "good" and "bad" result `server.log` that I hope is helpful.
[good_xpra_tcp-server.log](https://github.com/Xpra-org/xpra-html5/files/8757336/good_xpra_tcp-server.log)
[bad_xpra_tcp-server.log](https://github.com/Xpra-org/xpra-html5/files/8757337/bad_xpra_tcp-server.log)

**Comment by sbradnick** [comment]
I've rebuilt the rpm against a few revisions past ecc3a08 and it works [as I understand the usage] with:

4.5.2+git20220430.8aea0a5 [8aea0a5]
4.5.2+git20220501.0f51cab [0f51cab]

If it's of value to know, here's the xpra version: `xpra-4.4+git20220521.cb25967f8`

**Comment by sbradnick** [comment]
I was able to have success ([+]) with many commits up until 799330a ([x]) which is part of quite a batch on "May 16, 2022" ; so I'll start peeling back from there and see which is the last one that works.

[+] 4.5.2+git20220421.ecc3a08-1.1.noarch.rpm
[+] 4.5.2+git20220430.8aea0a5-0.noarch.rpm
[+] 4.5.2+git20220501.0f51cab-0.noarch.rpm
[+] 4.5.2+git20220503.1fe2469-0.noarch.rpm
[+] 4.5.2+git20220504.09d46e4-0.noarch.rpm
[+] 4.5.2+git20220510.3e2e896-0.noarch.rpm
[+] 5.0+git20220511.7908dab-0.noarch.rpm
[+] 5.0+git20220515.9512aca-0.noarch.rpm
[x] 5.0+git20220516.799330a-0.noarch.rpm
[x] 5.0+git20220523.59a8f90-0.noarch.rpm

With 799330a I was getting "WebSocket connection established" just sitting there, but a refresh of the browser (I was prompted for a p/w both times, which is good) gave me the familiar "Waiting for decode worker".

**Comment by sbradnick** [comment]
5.0+git20220516.fbb091f [fbb091f] seems to be the 1st time it flat out doesn't work.
5.0+git20220516.ced316f [ced316f] "works", but throws REPEATED errors like this:
```
2022-05-23 16:29:53,893 Warning: client decoding error:
2022-05-23 16:29:53,893  TypeError: Cannot read property 'request_redraw' of undefined
2022-05-23 16:29:53,920 client   1 TypeError: Cannot read property 'debug' of undefined
2022-05-23 16:29:53,921 client   1     at http://rpi4tw2.localdomain.net:14500/js/Client.js:1:77775
2022-05-23 16:29:53,922 client   1     at painted (http://rpi4tw2.localdomain.net:14500/js/Window.js:1:26010)
2022-05-23 16:29:53,923 client   1     at paint_bitmap (http://rpi4tw2.localdomain.net:14500/js/Window.js:1:26271)
2022-05-23 16:29:53,923 client   1     at XpraWindow.do_paint (http://rpi4tw2.localdomain.net:14500/js/Window.js:1:27002)
2022-05-23 16:29:53,924 client   1     at XpraWindow.paint (http://rpi4tw2.localdomain.net:14500/js/Window.js:1:24749)
2022-05-23 16:29:53,924 client   1     at XpraClient.do_process_draw (http://rpi4tw2.localdomain.net:14500/js/Client.js:1:77563)
2022-05-23 16:29:53,924 client   1     at Worker.<anonymous> (http://rpi4tw2.localdomain.net:14500/js/Client.js:1:10446)
2022-05-23 16:29:53,925 client   1 error painting webp TypeError: Cannot read property 'debug' of undefined
2022-05-23 16:29:53,925 client   1 TypeError: Cannot read property 'request_redraw' of undefined
2022-05-23 16:29:53,926 client   1     at http://rpi4tw2.localdomain.net:14500/js/Client.js:1:77734
2022-05-23 16:29:53,926 client   1     at paint_error (http://rpi4tw2.localdomain.net:14500/js/Window.js:1:26107)
2022-05-23 16:29:53,926 client   1     at XpraWindow.do_paint (http://rpi4tw2.localdomain.net:14500/js/Window.js:1:28314)
2022-05-23 16:29:53,927 client   1     at XpraWindow.paint (http://rpi4tw2.localdomain.net:14500/js/Window.js:1:24749)
2022-05-23 16:29:53,927 client   1     at XpraClient.do_process_draw (http://rpi4tw2.localdomain.net:14500/js/Client.js:1:77563)
2022-05-23 16:29:53,928 client   1     at Worker.<anonymous> (http://rpi4tw2.localdomain.net:14500/js/Client.js:1:10446)
```
The same goes for:
5.0+git20220516.9e26e8e [9e26e8e]
5.0+git20220516.ced316f [ced316f]
5.0+git20220516.8336b02 [8336b02]

I feel pretty confident in 5.0+git20220516.caf9971 [caf9971] being the last one to both work and not throw `TypeError: Cannot read property 'request_redraw' of undefined` errors.

**Comment by sbradnick** [comment]
My general flow for testing this is x86_64 (odroid sbc) => aarch64 (raspberry pi4) ; but now I'm using a Hyper-V install (just to rule something odd out between x86_64 <-> aarch64, though this package is a noarch).

Same issue(s) with xpra-html5-5.0+git20220525.0b674e0-0.noarch ...

**Comment by sbradnick** [comment]
Attaching 2 Developer Tools -> Console logs from the last working commit I have `caf9971` vs today's `0b674e0`. If there's something better I can supply, let me know :)

[success_172.30.65.144-1653498768661.log](https://github.com/Xpra-org/xpra-html5/files/8773250/success_172.30.65.144-1653498768661.log)
[failure_172.30.65.144-1653499165578.log](https://github.com/Xpra-org/xpra-html5/files/8773251/failure_172.30.65.144-1653499165578.log)

**Comment by totaam** [comment]
In both cases in tries to use the worker.
When it succeeds:
```
we can decode using a worker: Worker
Utilities.js:1 full list of supported encodings: Array(4)
```
When it fails:
```
waiting for decode worker to finish initializing
Decoder.js:1 Invalid asm.js: Unexpected token
Utilities.js:1 we can decode using a worker: Worker {onmessage: null, onerror: null}
```
But the latter uses authentication (`process challenge: xor`) - is this relevant?
Does authentication make a difference?

**Comment by sbradnick** [comment]
I believe the do_paint errors stopped occuring w/ 0b674e0, but I'll double check.

Also, `xpra start --start=xterm --bind-tcp=0.0.0.0:14500` works w/ 59a8f90, but I wouldn't think that's ideal (no authentication).

So neither:

`xpra start --start=xterm -d auth --bind-tcp=0.0.0.0:10000,auth=file:filename=password.txt`
nor
`xpra start --start=xterm -d auth --bind-tcp=0.0.0.0:14500,auth=sys`

give me a working session after caf9971.

**Comment by sbradnick** [comment]
I do get `do_paint` messages w/ qutebrowser (Backend: QtWebEngine 5.15.9, based on Chromium 87.0.4280.144), but not with Chromium (101.0.4951.67). Regardless, the xterm window can't be moved. Firefox (100.0.2) doesn't work at all (even without auth, as all 3 were tested). This is with 5.0+git20220526.39566a8 ...

All 3 browsers work as I'd expect via `xpra start --start=xterm -d auth --bind-tcp=0.0.0.0:14500,auth=sys`.

**Comment by totaam** [comment]
Authentication handler fix in 71e1ca340461d83aed3ab259f8bbeab67146087f.

**Comment by totaam** [comment]
> I do get do_paint messages

That's fixed in 3e504235afe7d756b7e6909afb52241f20010b35

**Comment by sbradnick** [comment]
Unfortunately nothing has changed for me using:

xpra-4.4+git20220523.d5a721cc4-1.1.x86_64
xpra-html5-5.0+git20220526.39566a8-0.noarch

I'm going to give it a go w/ today's latest xpra commit just to see and then re-install the last known working html5 commit to be sure.

Even though I can't move the xterm window, I can use it (even maximize/minimize it), when I close it w/ the top right X, the browser console window shows this:

![2022-05-31-113404_536x138_scrot](https://user-images.githubusercontent.com/84082961/171212857-5227a2f2-a580-45f5-98ed-7d9f91ddf27e.png)

Maybe that's normal ... ?

**Comment by totaam** [comment]
All the problems you report should be fixed in git master except for #188

**Comment by sbradnick** [comment]
Thanks for getting this fixed, all good now :+1:
